### PR TITLE
feat(connectors): HarborConnector skeleton — HTTP Basic + fingerprint + probe + registry v2 (G3.5-T7 #619)

### LINKED ISSUE #619: Task: G3.5-T7 HarborConnector skeleton — HttpConnector subclass + HTTP Basic (admin+robot) + fingerprint + probe + registry v2
Parent initiative: #368
Parent goal: #214
Depends-on: (none — substrate closed)

## Target

- Add: `backend/src/meho_backplane/connectors/harbor/{__init__.py,connector.py,session.py}`
- Add: `backend/tests/test_connectors_harbor_auth.py` (respx)
- Precedent: [`connectors/vmware_rest/`](../blob/main/backend/src/meho_backplane/connectors/vmware_rest/), [`connectors/adapters/http.py`](../blob/main/backend/src/meho_backplane/connectors/adapters/http.py).

## Context

Harbor is greenfield — the consumer has **no `harbor.sh` wrapper** (verified — not in the consumer `scripts/`). This is the first MEHO connector with no consumer wrapper to mirror; the contract reference is the Harbor 2.x OpenAPI spec, not a bash wrapper. Same `HttpConnector` shape as the vSphere precedent; auth is HTTP Basic (admin + robot accounts). This Task ships the skeleton; read ops are spec-ingested in #620 and the scoped-write robot lifecycle lands in #621.

## Current state

No `connectors/harbor/` package. Harbor targets unreachable through MEHO.

## Desired state

`HarborConnector(HttpConnector)`, `vendor="vmware"`, `product="harbor"`, `version="2.x"`, `impl_id="harbor-rest"` (Broadcom-acquired but the OSS Harbor identity remains), registered via `register_connector_v2(...)` in `__init__.py`.

Auth: **HTTP Basic** (admin or robot account); credentials from the target's Vault `secret_ref` via an injectable loader (mirror the vSphere loader shape; Basic sent per request via httpx `auth=(u,p)` — no session token). `auth_headers()` rejects non-`shared_service_account` `auth_model` with `NotImplementedError`.

`fingerprint()` → `GET /api/v2.0/systeminfo` → `FingerprintResult(vendor="vmware", product="harbor", version=..., build=..., extras={"auth_mode": ...})`; failure → `reachable=False` + structured `extras["error"]`. `probe()` → TCP + cert + `GET /api/v2.0/health` (Harbor's own composite healthcheck covering DB + redis + registry + jobservice). `execute()` = G0.6 dispatch shim (`connector_id="harbor-rest-2.x"`).

## Acceptance criteria

- [ ] `connectors/harbor/` exists; `register_connector_v2(product="harbor", version="2.x", impl_id="harbor-rest", cls=HarborConnector)` in `__init__.py`; registry-v2 test asserts the triple resolves.
- [ ] respx unit test covers: Basic auth (admin); Basic auth (robot-account username form); Vault loader missing key → clear `RuntimeError` naming target; per-target client isolation; `auth_model="per_user"` → `NotImplementedError`.
- [ ] `fingerprint()` against mocked `GET /api/v2.0/systeminfo` returns canonical shape; unreachable → `reachable=False` + `extras["error"]`.
- [ ] `probe()` maps `GET /api/v2.0/health` component statuses to a single `ok` + reason.
- [ ] `ruff` + `mypy` clean; `pytest -n auto backend/tests/test_connectors_harbor_auth.py backend/tests/test_connectors_registry_v2.py` passes.

## Out of scope

- Harbor read ops (project/repo/artifact lists) — #620.
- Harbor robot lifecycle (`robot.create`/`.delete`) + G6 credential_mint classifier — #621.
- Harbor image push/delete, replication, vuln-scan — deferred per #368.
- CLI / MCP review / onboarding doc — #622.

## References

- Precedent: [`connectors/vmware_rest/connector.py`](../blob/main/backend/src/meho_backplane/connectors/vmware_rest/connector.py), [`connectors/vmware_rest/session.py`](../blob/main/backend/src/meho_backplane/connectors/vmware_rest/session.py); `register_connector_v2` [`registry.py:108`](../blob/main/backend/src/meho_backplane/connectors/registry.py#L108).
- Pinned: httpx 0.28.1, tenacity 9.1.4, pydantic 2.13.4, respx 0.23.1.
- Harbor 2.x API / OpenAPI: https://goharbor.io/docs/2.11.0/build-customize-contribute/configure-swagger/
- Best-practices anchors (by pack name): **python_best_practices** — httpx pooling, Pydantic v2 schemas, Vault-loader injection for testability, retry discipline.

### PR DESCRIPTION
**Pull request by kr3s0**
## Summary

- Adds `connectors/harbor/` package (`__init__.py`, `connector.py`, `session.py`) with `HarborConnector(HttpConnector)` registered under `(product="harbor", version="2.x", impl_id="harbor-rest")` via `register_connector_v2`.
- Auth: HTTP Basic with verbatim username (admin or `robot$project+name` robot accounts), credentials cached per-target from injectable Vault loader — same pattern as SDDC Manager but without `sso_realm` suffix.
- `fingerprint()` hits `GET /api/v2.0/systeminfo`; splits `harbor_version` (e.g. `"v2.11.0-abc123"`) into separate `version` + `build`. `probe()` uses Harbor's dedicated `GET /api/v2.0/health` composite healthcheck (DB, redis, registry, jobservice) rather than delegating to `fingerprint()`.
- Ships `docs/codebase/connectors-harbor.md` and extends `test_connectors_registry_v2.py` with the harbor triple assertion (per AC).

## Test plan

- [x] `ruff check src/ tests/` — clean (pre-existing bind9 E501 unrelated)
- [x] `mypy src/ --ignore-missing-imports` — no issues in 189 source files
- [x] `pytest tests/test_connectors_harbor_auth.py` — 22 passed (all ACs covered)
- [x] `pytest tests/test_connectors_registry_v2.py::test_harbor_connector_registered_under_v2_triple` — passed
- [x] Basic auth admin account — correct `Authorization: Basic` header
- [x] Basic auth robot account (`robot$project+name` form) — passed through verbatim
- [x] Per-target credential isolation — loader called once per target
- [x] Vault loader missing key → `RuntimeError` naming target
- [x] `auth_model="per_user"` → `NotImplementedError` naming target + mode
- [x] `fingerprint()` canonical shape + version/build split
- [x] `fingerprint()` unreachable → `reachable=False` + `extras["error"]`
- [x] `probe()` all components healthy → `ok=True`
- [x] `probe()` unhealthy component → `ok=False` + reason listing names
- [x] `probe()` transport error → `ok=False` + reason

## Out of scope

- Harbor read ops (project/repo/artifact lists) — #620
- Harbor robot lifecycle (robot.create/delete) + G6 credential_mint classifier — #621
- CLI verbs + MCP review + real-container E2E — #622

Closes #619

## Summary by CodeRabbit

* **New Features**
  * Harbor 2.x container registry connector now available, featuring HTTP Basic authentication, per-target credential caching with isolation, system health checks, and version fingerprinting capabilities.
  * Connector is automatically registered and ready for immediate deployment across integrated systems.

* **Documentation**
  * Added Harbor 2.x connector documentation covering architecture, credential handling, authentication mechanisms, and API integration patterns.

[Review Change Stack](https://app.coderabbit.ai/change-stack/evoila/meho/pull/711?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: backend/src/meho_backplane/connectors/harbor/connector.py [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+        """Return the cached credentials for *target*, loading from Vault on first use.
+
+        The lock serialises concurrent first-use callers for the same target;
+        subsequent calls take the fast path under the same lock. The loaded
+        dict must contain ``"username"`` and ``"password"`` keys; missing
+        keys raise a :exc:`RuntimeError` naming the target and the missing
+        key so operators can identify a misconfigured Vault path.
+        """
+        async with self._creds_lock:
+            cached = self._creds_cache.get(target.name)
```
Comment (kr3s0) [inline backend/src/meho_backplane/connectors/harbor/connector.py:218]: **⚠️ Major** — credential cache keyed by `target.name` collides across tenants

**What:** `_creds_cache` uses `target.name` as the sole key. The connector is a global singleton (`get_or_create_connector_instance` in `_handler_resolve.py`). The DB enforces uniqueness on `(tenant_id, name)` — not `name` alone — so two tenants can each register a Harbor target named `"harbor-prod"` pointing to different instances with different credentials. When tenant-A's `"harbor-prod"` populates the cache, tenant-B's `"harbor-prod"` gets a cache hit on the wrong credentials.

**Why it matters:** Cross-tenant credential leak. Tenant-B authenticates to Harbor with tenant-A's service account until the connector instance is replaced. With G0.6's `get_or_create_connector_instance` returning a process-lifetime singleton, that cache entry can persist indefinitely.

**Required change:** Key on `target.secret_ref` (already present in `HarborTargetLike`) instead of `target.name`. The Vault path is the true identity of a credential set — two targets with different `secret_ref` values get separate cache entries even across tenants, while two targets that legitimately share a `secret_ref` correctly reuse cached credentials. When G0.3 (#224) lands the concrete `Target` model, migrate to `str(target.id)` for stronger isolation. Apply the same fix to `sddc_manager/connector.py` lines 192 and 205 (identical pattern).

```python
# connector.py _load_credentials — before / after
# Before
cached = self._creds_cache.get(target.name)
...
self._creds_cache[target.name] = raw

# After (interim — secret_ref is in HarborTargetLike today)
cached = self._creds_cache.get(target.secret_ref)
...
self._creds_cache[target.secret_ref] = raw
```

**Source:** `backend/src/meho_backplane/db/models.py` — `TargetORM` unique constraint on `(tenant_id, name)`; `backend/src/meho_backplane/operations/_handler_resolve.py` — `get_or_create_connector_instance` returns a class-level global.

**Review by kr3s0 — COMMENTED** [review]
## /review-pr-slim summary

**Anchor task:** #619 — G3.5-T7 HarborConnector skeleton (HttpConnector subclass + HTTP Basic + fingerprint + probe + v2 registry triple)
**Surface area:** backend (connectors)
**Best-practices packs audited:** python
**Verdict:** COMMENT

### Acceptance criteria
- [x] `HarborConnector` registered under `("harbor", "2.x", "harbor-rest")` v2 triple — `test_harbor_connector_registered_under_v2_triple` + `test_importing_package_registers_against_v2_registry` pass
- [x] HTTP Basic auth — admin username (`"admin"`) and robot-account form (`"robot$project+name"`) both sent verbatim — `test_auth_headers_sends_basic_auth_for_admin_account` + `test_auth_headers_sends_basic_auth_for_robot_account_username_form` verified header decode
- [x] Credentials cached per-target; loader called once — `test_auth_headers_reuses_cached_credentials_across_calls` asserts `call_count == 1`
- [x] `auth_model` gate: `None` / `SHARED_SERVICE_ACCOUNT` (value + enum member) accepted; `per_user`, `impersonation`, unknown raise `NotImplementedError` — `test_auth_headers_rejects_non_shared_service_account_modes` (parametrised) + `test_auth_headers_accepts_*` variants
- [x] `fingerprint()` via `GET /api/v2.0/systeminfo`; `harbor_version` split into `version` + `build` — canonical shape test + `test_fingerprint_version_without_build_hash` (no `-`)
- [x] `fingerprint()` transport / status failure → `reachable=False` + `extras["error"]` — `test_fingerprint_unreachable_returns_reachable_false_with_structured_error`
- [x] `probe()` via `GET /api/v2.0/health`; per-component mapping, unhealthy reason lists component names — `test_probe_returns_ok_true_*` + `test_probe_returns_ok_false_with_reason_*`
- [x] `aclose()` clears credential cache and httpx pool — `test_aclose_clears_credential_cache_and_pool`
- [x] `docs/codebase/connectors-harbor.md` created — verified present, accurate against implementation

### Project-board alignment
| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #619 (evoila-bosnia/meho-internal) | — (not tracked on project 19) | `Closes #619` in PR body | ✅ on merge | ✅ full | close on merge |

### Predicted vs. actual diff
- ✅ `connectors/harbor/__init__.py` — expected; registers v2 triple at import-time, same pattern as sddc_manager and nsx
- ✅ `connectors/harbor/connector.py` — expected; `HarborConnector(HttpConnector)` with auth, fingerprint, probe, execute shim
- ✅ `connectors/harbor/session.py` — expected; `HarborTargetLike` Protocol (no `sso_realm`), `HarborCredentialsLoader` type alias, `load_credentials_from_vault` stub
- ✅ `tests/test_connectors_harbor_auth.py` — expected; 22 tests covering full AC surface + respx mocks
- ✅ `tests/test_connectors_registry_v2.py` — expected; harbor triple assertion appended
- ✅ `docs/codebase/connectors-harbor.md` — expected per task + Phase 2.5 doc gap check

### Integration verification
- [x] `ruff check src/ tests/` — clean (pre-existing bind9 E501 not introduced here)
- [x] `ruff format --check` — clean
- [x] `mypy src/ --ignore-missing-imports` — no issues (189 source files)
- [x] `pytest tests/test_connectors_harbor_auth.py` — 22 passed
- [x] `pytest tests/test_connectors_registry_v2.py::test_harbor_connector_registered_under_v2_triple` — passed
- ⚠️ not run: CI checks (Python lane, TruffleHog, Semgrep, Docker build) still QUEUED at post time — local gates match what CI runs; pre-existing `test_sddc_manager_connector_registered_under_v2_triple` failure confirmed via `git stash` to predate this PR
- ⚠️ not run: real-container E2E — out of scope per task (arrives in #622)

### Findings (inline — each is a merge blocker per team policy)
- ⚠️ 1 major — see inline at `connector.py:218` (credential cache key)

### Adjacent improvements (non-blocking, not posted inline)
- `sddc_manager/connector.py` lines 192 and 205 carry the identical cross-tenant cache key pattern; the inline comment's `secret_ref` fix should apply to both connectors in the same commit or a coordinated follow-up issue so the attack surface is closed uniformly. Filing a single issue to track both is the right scope.
- `docs/codebase/connectors-harbor.md` → "Known issues" could explicitly note the interim `secret_ref` cache key and point to G0.3 (#224) as the permanent fix, so the migration PR doesn't have to rediscover the rationale.

### Docs
- [x] `docs/codebase/connectors-harbor.md` — created; all sections present (Overview, Key types, Control flow, Dependencies, Known issues, References). Content verified accurate against implementation.

### What worked well
- Protocol-based `HarborTargetLike` with injectable `HarborCredentialsLoader` is the right shape: unit tests stay clean, the real Vault loader is cleanly deferred to G0.3, and the constructor default wires up correctly without extra configuration overhead.
- The dedicated `GET /api/v2.0/health` probe path (rather than delegating to `fingerprint()`) is the correct architecture for Harbor — the health endpoint covers subsystem state (redis, jobservice, registry) that `systeminfo` doesn't expose.
- `_parse_harbor_version`'s split-on-first-dash handles both `"v2.11.0-abc1234"` and bare `"v2.11.0"` robustly; the `or None` guards on empty strings after the split are correct edge-case handling.
- 22-test suite achieves near-exhaustive AC coverage including parametrised auth-model rejection, robot-account username passthrough, per-target credential isolation, and all three probe/fingerprint outcome paths.

File: backend/src/meho_backplane/connectors/harbor/connector.py (hunk 1 above)
Comment (kr3s0) [inline backend/src/meho_backplane/connectors/harbor/connector.py:218]: **⚠️ Major** — credential cache keyed by `target.name` collides across tenants

**What:** `_creds_cache` uses `target.name` as the sole key. The connector is a global singleton (`get_or_create_connector_instance` in `_handler_resolve.py`). The DB enforces uniqueness on `(tenant_id, name)` — not `name` alone — so two tenants can each register a Harbor target named `"harbor-prod"` pointing to different instances with different credentials. When tenant-A's `"harbor-prod"` populates the cache, tenant-B's `"harbor-prod"` gets a cache hit on the wrong credentials.

**Why it matters:** Cross-tenant credential leak. Tenant-B authenticates to Harbor with tenant-A's service account until the connector instance is replaced. With G0.6's `get_or_create_connector_instance` returning a process-lifetime singleton, that cache entry can persist indefinitely.

**Required change:** Key on `target.secret_ref` (already present in `HarborTargetLike`) instead of `target.name`. The Vault path is the true identity of a credential set — two targets with different `secret_ref` values get separate cache entries even across tenants, while two targets that legitimately share a `secret_ref` correctly reuse cached credentials. When G0.3 (#224) lands the concrete `Target` model, migrate to `str(target.id)` for stronger isolation. Apply the same fix to `sddc_manager/connector.py` lines 192 and 205 (identical pattern).

```python
# connector.py _load_credentials — before / after
# Before
cached = self._creds_cache.get(target.name)
...
self._creds_cache[target.name] = raw

# After (interim — secret_ref is in HarborTargetLike today)
cached = self._creds_cache.get(target.secret_ref)
...
self._creds_cache[target.secret_ref] = raw
```

**Source:** `backend/src/meho_backplane/db/models.py` — `TargetORM` unique constraint on `(tenant_id, name)`; `backend/src/meho_backplane/operations/_handler_resolve.py` — `get_or_create_connector_instance` returns a class-level global.

**Review by zdamir — APPROVED** [review]
